File: survey/drivers/sonatype/package_url_java/src/main/java/format.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.TreeMap;

import org.sonatype.goodies.packageurl.InvalidException;
import org.sonatype.goodies.packageurl.PackageUrl;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.DatabindException;
import com.fasterxml.jackson.databind.ObjectMapper;

public class format {
    public static void main(String[] args) throws IOException, DatabindException {
        BufferedReader s = new BufferedReader(new InputStreamReader(System.in));
        ObjectMapper mapper = new ObjectMapper();
        mapper.configure(JsonGenerator.Feature.AUTO_CLOSE_TARGET, false);
        while (true) {
            String line = s.readLine();
            if (line == null) {
                break;
            }
            line = line.trim();
            if (line.isEmpty()) {
                continue;
            }
            try {
                Parts parts = mapper.readValue(line, Parts.class);
                TreeMap<String, String> qualifiers = null;
                if (parts.qualifiers != null) {
                    qualifiers = new TreeMap<String, String>(parts.qualifiers);
                }
                PackageUrl purl = PackageUrl.builder().type(parts.type).namespace(parts.namespace).name(parts.name).version(parts.version).qualifiers(qualifiers).subpath(parts.subpath).build();
                System.out.println(purl);
            } catch (InvalidException e) {
                Error error = new Error();
                error.error = e.toString();
                mapper.writeValue(System.out, error);
                System.out.println();
            }
        }
    }
}

File: survey/drivers/althonos/packageurl_rs/src/format.rs
use std::io::{self, Write};

use packageurl::{Error, PackageUrl};

mod json;

fn main() -> io::Result<()> {
    let mut line = String::new();
    loop {
        line.clear();
        io::stdin().read_line(&mut line)?;
        if line.is_empty() {
            break;
        }
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let parts: json::Parts = match serde_json::from_str(line) {
            Ok(parts) => parts,
            Err(error) => {
                let mut out = io::stdout().lock();
                serde_json::to_writer(
                    &mut out,
                    &json::Error {
                        error: error.to_string(),
                    },
                )
                .unwrap();
                writeln!(out)?;
                out.flush()?;
                continue;
            }
        };
        let mut out = io::stdout().lock();
        match build(&parts) {
            Ok(purl) => writeln!(out, "{purl}")?,
            Err(error) => {
                serde_json::to_writer(
                    &mut out,
                    &json::Error {
                        error: format!("{error:?}"),
                    },
                )
                .unwrap();
                writeln!(out)?;
            }
        }
        out.flush()?;
    }

    Ok(())
}

fn build(parts: &json::Parts) -> Result<PackageUrl, Error> {
    let mut builder = PackageUrl::new(&parts.r#type, &parts.name)?;
    if let Some(namespace) = &parts.namespace {
        builder.with_namespace(namespace);
    }
    if let Some(version) = &parts.version {
        builder.with_version(version);
    }
    for (k, v) in parts.qualifiers.iter() {
        builder.add_qualifier(k, v)?;
    }
    if let Some(subpath) = &parts.subpath {
        builder.with_subpath(subpath)?;
    }
    Ok(builder)
}
